### Bibliotecas usadas no projeto:

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

### Carregando os dados. Dados retirados de: https://dadosabertos.aneel.gov.br/dataset/relacao-de-empreendimentos-de-geracao-distribuida

In [2]:
df = pd.read_csv("dados.csv")

### Um resumo dos dados, as primeiras linhas:

In [3]:
df.head()

_id DatGeracaoConjuntoDados CodGeracaoDistribuida MdaAreaArranjo  \
0    1     2025-06-30T00:00:00                   NaN           5,82   
1    2     2025-06-30T00:00:00                   NaN           2,47   
2    3     2025-06-30T00:00:00                   NaN          14,89   
3    4     2025-06-30T00:00:00                   NaN           4,23   
4    5     2025-06-30T00:00:00                   NaN           4,55   

  MdaPotenciaInstalada NomFabricanteModulo NomFabricanteInversor  \
0                 6,00              Não há                Não há   
1                 5,00              Não há                Não há   
2                 3,00              Não há                Não há   
3                 5,00              Não há                Não há   
4                 6,00              Não há                Não há   

            DatConexao MdaPotenciaModulos MdaPotenciaInversores  QtdModulos  \
0  2024-02-01T00:00:00               6,00                  6,00         852   
1  2024-02-01T00:00:00               5,00                  5,00          75   
2  2024-02-01T00:00:00               3,00                  3,00         626   
3  2024-02-01T00:00:00               5,00                  5,00         424   
4  2024-02-01T00:00:00               6,00                  6,00         495   

  NomModeloModulo NomModeloInversor  
0          Não há            Não há  
1          Não há            Não há  
2          Não há            Não há  
3          Não há            Não há  
4          Não há            Não há

### Modificando o nome das colunas e alterando "Não ha" por NaN, para melhor compreensão:

In [4]:
df.replace("Não há", np.nan, inplace=True)
df.rename(columns={'DatGeracaoConjuntoDados':'Data_Geracao_Dados'}, inplace=True)
df.rename(columns={'CodGeracaoDistribuida':'Codigo_Unidade_Geradora'}, inplace=True)
df.rename(columns={'MdaAreaArranjo':'Total_Area_Arranjo'}, inplace=True)
df.rename(columns={'MdaPotenciaInstalada':'Potencia_Instalada'}, inplace=True)
df.rename(columns={'NomFabricanteModulo':'Fabricante_Modulo'}, inplace=True)
df.rename(columns={'NomFabricanteInversor':'Fabricante_Inversor'}, inplace=True)
df.rename(columns={'DatConexao':'Data_conexao_Unidade'}, inplace=True)
df.rename(columns={'MdaPotenciaModulos':'Potencia_Modulos'}, inplace=True)
df.rename(columns={'MdaPotenciaInversores':'Potencia_Inversor'}, inplace=True)
df.rename(columns={'QtdModulos':'Quantidade_Modulos'}, inplace=True)
df.rename(columns={'NomModeloModulo':'Nome_Modelo_Modulo'}, inplace=True)
df.rename(columns={'NomModeloInversor':'Nome_Modelo_inversor'}, inplace=True)

# Dicionário de Dados — Geração Distribuída Fotovoltaica

| **Nome do Campo**              | **Tipo do Dado**         | **Tamanho** | **Descrição** |
|-------------------------------|--------------------------|-------------|---------------|
| `_id`                         | Numérico                 | —           | Identificador único da linha no dataset. |
| `Data_Geracao_Dados`         | Data Simples             | —           | Data do processamento de carga automática no momento da geração para publicação do conjunto de dados abertos. |
| `Codigo_Unidade_Geradora`    | Cadeia de caracteres     | 21          | Código identificador da unidade geradora (UG) no sistema da distribuidora. |
| `Total_Area_Arranjo`         | Numérico                 | 8,2         | Área total dos arranjos fotovoltaicos em metros quadrados (m²). |
| `Potencia_Instalada`         | Numérico                 | 6,2         | Potência instalada da unidade geradora, segundo o projeto aprovado, em quilowatts (kW). |
| `Fabricante_Modulo`          | Cadeia de caracteres     | 100         | Nome do fabricante dos módulos fotovoltaicos. |
| `Fabricante_Inversor`        | Cadeia de caracteres     | 100         | Nome do fabricante dos inversores. |
| `Data_conexao_Unidade`       | Data Simples             | —           | Data em que a unidade geradora foi conectada à rede da distribuidora. |
| `Potencia_Modulos`           | Numérico                 | 6,2         | Soma das potências nominais dos módulos instalados, em kW. |
| `Potencia_Inversor`          | Numérico                 | 6,2         | Soma das potências nominais dos inversores, em kW. |
| `Quantidade_Modulos`         | Numérico                 | 10          | Número total de módulos fotovoltaicos instalados. |
| `Nome_Modelo_Modulo`         | Cadeia de caracteres     | 100         | Modelo técnico ou código comercial do módulo utilizado. |
| `Nome_Modelo_inversor`       | Cadeia de caracteres     | 100         | Modelo técnico ou código comercial do inversor utilizado. |

### Coluna 00: _id -> Contagem de linhas;
### Coluna 01: Data_Geracao_Dados -> Data de processamento e publicação dos dados na plataforma que disponibilizou (publicamente) os dados;
### Coluna 02: Codigo_Unidade_Geradora -> Um identificador único (como se fosse um "cpf" da usida);
### Coluna 03: Total_Area_Arranjo -> Área total dos módulos (em $m^{2}$), ou seja, a área ocupada pelos painéis fotovotáicos;
### Coluna 04: Potencia_Instalada -> Potência total nominal instalada (em $kW$), que a unidade é capaz de gerar, segundo o projeto;
### Coluna 05: Fabricante_Modulo -> Nome do fabricante dos painéis, nome da empresa que fabricou os módulos solares;
### Coluna 06: Fabricante_Inversor -> Nome do fabricante dos inversores, nome da empresa que fabricou os inversores (equipamente que converte CC para CA);
### Coluna 07: Data_conexao_Unidade -> Data de conexão à rede, em que a unidade começou a injetar energia na rede da distribuidora (reflete o início da operação efetiva);
### Coluna 08: Potencia_Modulos -> Potência somada dos módulos (em $kW$), soma das potências nominais dos painéis;
### Coluna 09: Potencia_Inverso -> Potência somada dos Inversores (em $kW$), soma das potências nominais dos inversores conectados;
### Coluna 10: Quantidade_Modulos -> Quantidade total de painéis instalados, número de módulos instalados;
### Coluna 11: Nome_Modelo_Modulo -> Nome técnico do modelo do painel fotovotaico;
### Coluna 12: Nome_Modelo_Inversor -> Nome técnico do modelo do Inversor.

### Ficamos agora com todos os dados faltantes iguais a NaN:

In [5]:
df.head()

_id   Data_Geracao_Dados Codigo_Unidade_Geradora Total_Area_Arranjo  \
0    1  2025-06-30T00:00:00                     NaN               5,82   
1    2  2025-06-30T00:00:00                     NaN               2,47   
2    3  2025-06-30T00:00:00                     NaN              14,89   
3    4  2025-06-30T00:00:00                     NaN               4,23   
4    5  2025-06-30T00:00:00                     NaN               4,55   

  Potencia_Instalada Fabricante_Modulo Fabricante_Inversor  \
0               6,00               NaN                 NaN   
1               5,00               NaN                 NaN   
2               3,00               NaN                 NaN   
3               5,00               NaN                 NaN   
4               6,00               NaN                 NaN   

  Data_conexao_Unidade Potencia_Modulos Potencia_Inversor  Quantidade_Modulos  \
0  2024-02-01T00:00:00             6,00              6,00                 852   
1  2024-02-01T00:00:00             5,00              5,00                  75   
2  2024-02-01T00:00:00             3,00              3,00                 626   
3  2024-02-01T00:00:00             5,00              5,00                 424   
4  2024-02-01T00:00:00             6,00              6,00                 495   

  Nome_Modelo_Modulo Nome_Modelo_inversor  
0                NaN                  NaN  
1                NaN                  NaN  
2                NaN                  NaN  
3                NaN                  NaN  
4                NaN                  NaN

### Porém, temos muitos dados Faltantes. Podemos filtrar o Data Frame (df) e deixar apenas os dados completos. Vamos primeiro usar len(df), onde saberemos a quantidade de linhas:

In [6]:
len(df)

3643608

### Com essa informação, poderemos calcular a porcentagem de linhas filtradas do df. Vamos filtrar os NaN do df:

In [7]:
df = df.dropna()

### Podemos visualizar algumas linhas do df filtrado:

In [8]:
df.head(20)

_id   Data_Geracao_Dados Codigo_Unidade_Geradora Total_Area_Arranjo  \
38   39  2025-06-30T00:00:00       GD.AC.000.000.590             236,00   
39   40  2025-06-30T00:00:00       GD.AC.000.001.540              12,88   
40   41  2025-06-30T00:00:00       GD.AC.000.001.552              13,04   
41   42  2025-06-30T00:00:00       GD.AC.000.009.608              38,40   
42   43  2025-06-30T00:00:00       GD.AC.000.017.465              36,00   
43   44  2025-06-30T00:00:00       GD.AC.000.017.466              64,00   
44   45  2025-06-30T00:00:00       GD.AC.000.017.513              25,00   
45   46  2025-06-30T00:00:00       GD.AC.000.023.039              64,00   
46   47  2025-06-30T00:00:00       GD.AC.000.023.041              36,00   
47   48  2025-06-30T00:00:00       GD.AC.000.023.042              40,00   
48   49  2025-06-30T00:00:00       GD.AC.000.023.043              48,00   
49   50  2025-06-30T00:00:00       GD.AC.000.024.832             312,00   
50   51  2025-06-30T00:00:00       GD.AC.000.024.973              60,00   
51   52  2025-06-30T00:00:00       GD.AC.000.025.059              42,80   
52   53  2025-06-30T00:00:00       GD.AC.000.025.061              25,00   
53   54  2025-06-30T00:00:00       GD.AC.000.026.897              63,34   
54   55  2025-06-30T00:00:00       GD.AC.000.028.769              25,00   
55   56  2025-06-30T00:00:00       GD.AC.000.028.793              55,00   
56   57  2025-06-30T00:00:00       GD.AC.000.032.431              30,00   
57   58  2025-06-30T00:00:00       GD.AC.000.032.844               2,08   

   Potencia_Instalada Fabricante_Modulo Fabricante_Inversor  \
38              32,50    Canadian Solar             Fronius   
39               2,00    Canadian Solar            Santerno   
40               2,00        AV PROJECT             Fronius   
41               5,00    Canadian Solar             Fronius   
42               5,00    CANADIAN SOLAR             FRONIUS   
43               8,20    CANADIAN SOLAR             FRONIUS   
44               3,00          CANADIAN                 ABB   
45              10,40    CANADIAN SOLAR             REFUSOL   
46               5,00    CANADIAN SOLAR             FRONIUS   
47               6,50    CANADIAN SOLAR             FRONIUS   
48               7,80    CANADIAN SOLAR             FRONIUS   
49              51,48    CANADIAN SOLAR             FRONIUS   
50               9,45          CANADIAN             FRONIUS   
51               7,26    CANADIAN SOLAR             FRONIUS   
52               3,65          CANADIAN            CANADIAN   
53              10,08    CANADIAN SOLAR             FRONIUS   
54               3,00    CANADIAN SOLAR                 ABB   
55               5,15    CANADIAN SOLAR                 ABB   
56               4,62         J A SOLAR                 ABB   
57              34,32    canadian Solar             FRONIUS   

   Data_conexao_Unidade Potencia_Modulos Potencia_Inversor  \
38  2016-08-31T00:00:00            36,80             32,50   
39  2016-10-31T00:00:00             2,00              2,00   
40  2016-08-25T00:00:00             2,00              2,00   
41  2017-03-09T00:00:00             5,00              5,00   
42  2017-07-29T00:00:00             5,00              5,00   
43  2017-10-05T00:00:00            10,40              8,20   
44  2017-11-22T00:00:00             3,00              3,00   
45  2017-12-01T00:00:00            10,40             10,40   
46  2018-01-03T00:00:00             5,85              5,00   
47  2017-12-26T00:00:00             6,50              6,50   
48  2017-12-28T00:00:00             7,80              7,80   
49  2018-03-17T00:00:00            51,48             51,48   
50  2018-03-21T00:00:00             9,45              9,45   
51  2017-12-04T00:00:00             7,26              7,26   
52  2018-02-19T00:00:00             3,65              3,65   
53  2017-12-21T00:00:00            10,08             10,08   
54  2018-02-19T00:00:00             3,00          

### Foram removidas um total de 54831 linhas de um total de 3643608 linhas.

In [9]:
len(df)

3588777

In [10]:
dif_linhas = ((3643608-3588777)*100)/3643608
print(dif_linhas)

1.5048545288077093


### A diferença de um data frame para o outro (após filtar as linhas com possuem NaN - não possuem valor, é muito pequena, então vamos trabalhar apenas com dados completos:

$$
    DifLinhas = \frac{(3643608-3588777)}{3643608}*100 = \frac{54831}{3643608}*100 \simeq 1.5048545288077093 \%
$$

### Resetamos a primeira coluna para os índices iniciem a contagem de 0 até o total de linhas (menos 1):

In [11]:
df = df.reset_index(drop=True)
df['_id'] = df.index

In [12]:
df.head(20)

_id   Data_Geracao_Dados Codigo_Unidade_Geradora Total_Area_Arranjo  \
0     0  2025-06-30T00:00:00       GD.AC.000.000.590             236,00   
1     1  2025-06-30T00:00:00       GD.AC.000.001.540              12,88   
2     2  2025-06-30T00:00:00       GD.AC.000.001.552              13,04   
3     3  2025-06-30T00:00:00       GD.AC.000.009.608              38,40   
4     4  2025-06-30T00:00:00       GD.AC.000.017.465              36,00   
5     5  2025-06-30T00:00:00       GD.AC.000.017.466              64,00   
6     6  2025-06-30T00:00:00       GD.AC.000.017.513              25,00   
7     7  2025-06-30T00:00:00       GD.AC.000.023.039              64,00   
8     8  2025-06-30T00:00:00       GD.AC.000.023.041              36,00   
9     9  2025-06-30T00:00:00       GD.AC.000.023.042              40,00   
10   10  2025-06-30T00:00:00       GD.AC.000.023.043              48,00   
11   11  2025-06-30T00:00:00       GD.AC.000.024.832             312,00   
12   12  2025-06-30T00:00:00       GD.AC.000.024.973              60,00   
13   13  2025-06-30T00:00:00       GD.AC.000.025.059              42,80   
14   14  2025-06-30T00:00:00       GD.AC.000.025.061              25,00   
15   15  2025-06-30T00:00:00       GD.AC.000.026.897              63,34   
16   16  2025-06-30T00:00:00       GD.AC.000.028.769              25,00   
17   17  2025-06-30T00:00:00       GD.AC.000.028.793              55,00   
18   18  2025-06-30T00:00:00       GD.AC.000.032.431              30,00   
19   19  2025-06-30T00:00:00       GD.AC.000.032.844               2,08   

   Potencia_Instalada Fabricante_Modulo Fabricante_Inversor  \
0               32,50    Canadian Solar             Fronius   
1                2,00    Canadian Solar            Santerno   
2                2,00        AV PROJECT             Fronius   
3                5,00    Canadian Solar             Fronius   
4                5,00    CANADIAN SOLAR             FRONIUS   
5                8,20    CANADIAN SOLAR             FRONIUS   
6                3,00          CANADIAN                 ABB   
7               10,40    CANADIAN SOLAR             REFUSOL   
8                5,00    CANADIAN SOLAR             FRONIUS   
9                6,50    CANADIAN SOLAR             FRONIUS   
10               7,80    CANADIAN SOLAR             FRONIUS   
11              51,48    CANADIAN SOLAR             FRONIUS   
12               9,45          CANADIAN             FRONIUS   
13               7,26    CANADIAN SOLAR             FRONIUS   
14               3,65          CANADIAN            CANADIAN   
15              10,08    CANADIAN SOLAR             FRONIUS   
16               3,00    CANADIAN SOLAR                 ABB   
17               5,15    CANADIAN SOLAR                 ABB   
18               4,62         J A SOLAR                 ABB   
19              34,32    canadian Solar             FRONIUS   

   Data_conexao_Unidade Potencia_Modulos Potencia_Inversor  \
0   2016-08-31T00:00:00            36,80             32,50   
1   2016-10-31T00:00:00             2,00              2,00   
2   2016-08-25T00:00:00             2,00              2,00   
3   2017-03-09T00:00:00             5,00              5,00   
4   2017-07-29T00:00:00             5,00              5,00   
5   2017-10-05T00:00:00            10,40              8,20   
6   2017-11-22T00:00:00             3,00              3,00   
7   2017-12-01T00:00:00            10,40             10,40   
8   2018-01-03T00:00:00             5,85              5,00   
9   2017-12-26T00:00:00             6,50              6,50   
10  2017-12-28T00:00:00             7,80              7,80   
11  2018-03-17T00:00:00            51,48             51,48   
12  2018-03-21T00:00:00             9,45              9,45   
13  2017-12-04T00:00:00             7,26              7,26   
14  2018-02-19T00:00:00             3,65              3,65   
15  2017-12-21T00:00:00            10,08             10,08   
16  2018-02-19T00:00:00             3,00          

### Vamos olhar algumas estatísticas básicas agora:

In [13]:
df.describe(include='all')

_id   Data_Geracao_Dados Codigo_Unidade_Geradora  \
count   3.588777e+06              3588777                 3588777   
unique           NaN                    1                 3588777   
top              NaN  2025-06-30T00:00:00       GD.AC.000.000.590   
freq             NaN              3588777                       1   
mean    1.794388e+06                  NaN                     NaN   
std     1.035991e+06                  NaN                     NaN   
min     0.000000e+00                  NaN                     NaN   
25%     8.971940e+05                  NaN                     NaN   
50%     1.794388e+06                  NaN                     NaN   
75%     2.691582e+06                  NaN                     NaN   
max     3.588776e+06                  NaN                     NaN   

       Total_Area_Arranjo Potencia_Instalada Fabricante_Modulo  \
count             3588777            3588777           3588777   
unique              36063               6495             58610   
top                 20,00               5,00             JINKO   
freq               199932             446950            150125   
mean                  NaN                NaN               NaN   
std                   NaN                NaN               NaN   
min                   NaN                NaN               NaN   
25%                   NaN                NaN               NaN   
50%                   NaN                NaN               NaN   
75%                   NaN                NaN               NaN   
max                   NaN                NaN               NaN   

       Fabricante_Inversor Data_conexao_Unidade Potencia_Modulos  \
count              3588777              3588777          3588777   
unique               43745                 3914            17558   
top                GROWATT  2022-12-29T00:00:00             4,40   
freq                426745                 5322            80178   
mean                   NaN                  NaN              NaN   
std                    NaN                  NaN              NaN   
min                    NaN                  NaN              NaN   
25%                    NaN                  NaN              NaN   
50%                    NaN                  NaN              NaN   
75%                    NaN                  NaN              NaN   
max                    NaN                  NaN              NaN   

       Potencia_Inversor  Quantidade_Modulos Nome_Modelo_Modulo  \
count            3588777        3.588777e+06            3588777   
unique              4735                 NaN             298344   
top                 5,00                 NaN    JAM72S30-550/MR   
freq              684798                 NaN              44158   
mean                 NaN        5.580174e+01                NaN   
std                  NaN        8.479818e+03                NaN   
min                  NaN        0.000000e+00                NaN   
25%                  NaN        9.000000e+00                NaN   
50%                  NaN        1.300000e+01                NaN   
75%                  NaN        2.200000e+01                NaN   
max                  NaN        8.699437e+06                NaN   

       Nome_Modelo_inversor  
count               3588777  
unique               228031  
top        SUN2000G3-US-220  
freq                  71270  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN